### ISSUE TITLE
PIP-124: Pulsar Client Shared  State API (draft)

### ISSUE THREAD

**Issue by eolivelli** [apache/pulsar#13490]
## Motivation

Explain why this change is needed, what benefits it would bring to Apache Pulsar
and what problem it's trying to solve.

## Goal

Define the scope of this proposal. Given the motivation stated above, what are
the problems that this proposal is addressing and what other items will be
considering out of scope, perhaps to be left to a different PIP.

## API Changes

Illustrate all the proposed changes to the API or wire protocol, with examples
of all the newly added classes/methods, including Javadoc.

## Implementation

This should be a detailed description of all the changes that are
expected to be made. It should be detailed enough that any developer that is
familiar with Pulsar internals would be able to understand all the parts of the
code changes for this proposal.

This should also serve as documentation for any person that is trying to
understand or debug the behavior of a certain feature.

## Reject Alternatives

If there are alternatives that were already considered by the authors or,
after the discussion, by the community, and were rejected, please list them
here along with the reason why they were rejected.

**Comment by mattisonchao** [comment]
Could i help with this PIP ?

**Comment by RobertIndie** [comment]
This PIP number is duplicated with https://github.com/apache/pulsar/issues/13408

**Comment by eolivelli** [comment]
@mattisonchao yes you could, but most of the code is already in the attached github repo.

**Comment by eolivelli** [comment]
we talked in the Community meeting about this PIP.

The main concern from @merlimat  is the exposing a "put" operation will give a false illusion that you have something like a Map, but actually every "write" operation will be "slow" because we need to acquire the Producer lock.

@merlimat suggested to split this into two distinct interfaces: a API for readers and one for writers, possibly extending the work done on PIP-104 TableViews (https://github.com/apache/pulsar/issues/12356).

I am fine with extending PIP-104 in that direction (to add a writer side of the TableView), but I believe that the API proposed here is going to give a API to achieve the goal of having a generic "Java object" as share state, by letting the developer of the Shared State Object deal with the internal representation.
Using the TableView API will forces users to represent their Java object as a set of properties or a single object with one fixed key. It is feasible, but not straightforward from my point of view.
I really would like to mimic Pravega.io StateSynchronizer support, that does not force you to see your topic as a "table".

@merlimat 
I will work in parallel on extending the TableView API to add writer support, I am sure it will bring a big value to the users.
But I still thing that we should add this "Recipe" to the Pulsar Adapters repository

@merlimat  WDTY ?
